Mixpanel/Mixpanel.m

```diff
@@ -693,6 +702,39 @@ - (NSURLRequest *)apiRequestWithEndpoint:(NSString *)endpoint andBody:(NSString
     return request;
 }
 
+- (void)handleNetworkResponse:(NSHTTPURLResponse *)response withError:(NSError *)error
+{
+    NSNumber *retryTime = response.allHeaderFields[@"Retry-After"];
+    
+    BOOL was5XX = (response.statusCode >= 500 && response.statusCode < 600);
+    if (was5XX) {
+        self.networkConsecutiveFailures++;
+    } else {
+        self.networkConsecutiveFailures = 0;
+    }
+    
+    if (self.networkConsecutiveFailures > 1) {
+        // Exponential backoff
+        retryTime = [self retryBackOffTime];
+    }
+    
+    if (!retryTime) {
+        BOOL wasTimeout = (error.code == NSURLErrorTimedOut);
+        if (wasTimeout || was5XX) {
+            // Prevent network requests for the next minute
+            retryTime = @(60);
+        }
+    }
+    
+    self.networkRequestsAllowedAfterTime = [[NSDate dateWithTimeIntervalSinceNow:retryTime.doubleValue] timeIntervalSince1970];
+}
+
+- (NSNumber *)retryBackOffTime
+{
+    const NSUInteger retryInterval = (arc4random() % 30) + 30;
+    return @(pow(retryInterval, self.networkConsecutiveFailures));
```

Comment:
in fact you could simplify this whole business to

    retryTime = min(max(60, [self retryBackOffTime]), 600)